Introduce broader scope for custom rules

sesponda opened the pull request:
This (preliminary) addresses https://github.com/StyraInc/regal/issues/282

I'd recommend reviewing it commit by commit.

1) First commit: introduced the change in the simplest possible way. Unit and e2e tests still green.
2) Second change: the added function (`collectAggregates`) duplicates the file iteration logic happening goroutines in `lintWithRegoRules`. The common logic is extracted into a new third function, `evalAndCollect`. This makes the diff a little messy, being the reason why it's in another comment.

Next actions: if this implementation is OK, I'll be happy to add more commits with unit tests and documentation updates (some expectations/guidance welcome)

anderseknert commented on pkg/linter/linter.go:
```diff
@@ -369,7 +374,7 @@ func (l Linter) paramsToRulesConfig() map[string]any {
 }
 
-func (l Linter) prepareRegoArgs() []func(*rego.Rego) {
+func (l Linter) prepareRegoArgs(query ast.Body) []func(*rego.Rego) {
```
This looks unused?

sesponda replied:
It's used by `lintWithRegoRules` (where query is `violations = data.regal.main.report`) and by `collectAggregates` (where query is `aggregates = data.regal.main.aggregate`)

anderseknert replied:
Ah, all good!

anderseknert commented on pkg/linter/linter.go:
```diff
@@ -676,3 +662,103 @@ func (l Linter) getBundleByName(name string) (*bundle.Bundle, error) {
+// run sequentially via a mutex.
+func (l Linter) evalAndCollect(ctx context.Context, input rules.Input, query rego.PreparedEvalQuery,
+	queryInputBuilder func(name string, content string, module *ast.Module) (map[string]any, error),
+	reportCollector func(report report.Report),
```
Could we perhaps define types for these callback functions to help keep the args list readable?

anderseknert commented on pkg/linter/linter.go:
```diff
@@ -206,7 +206,12 @@ func (l Linter) Lint(ctx context.Context) (report.Report, error) {
 	aggregate.Violations = append(aggregate.Violations, goReport.Violations...)
 
-	regoReport, err := l.lintWithRegoRules(ctx, input)
+	aggregates, err := l.collectAggregates(ctx, input)
```
I think guarding this with a:

```go
	if len(input.Modules) > 1 {
		// ...
	}
```
Would be enough to avoid having these rules called when only a single file is to be linted.

anderseknert reviewed (commented):
Looks like a great start! Some tests would help me understand exactly how this behaves, but I reckon that'll be the next step. Thanks a lot for looking into this! 🚀

charlieegan3 commented on pkg/linter/linter.go:
```diff
@@ -206,7 +206,12 @@ func (l Linter) Lint(ctx context.Context) (report.Report, error) {
+		return report.Report{}, fmt.Errorf("failed to collect aggregates using Rego rules: %w", err)
+	}
+
+	regoReport, err := l.lintWithRegoRules(ctx, input, aggregates)
```
question: I'm missing something I think. What I see is that the aggregates rules are collected (L209) before being passed to `l.lintWithRegoRules`. This in turn passes them to `evalAndCollect`. These are then all collected onto the report.

What wasn't clear to me in this implementation was when aggregate rules were evaluated on the results of other rules.

In my head, I can't see where step 4 from https://github.com/StyraInc/regal/issues/282#issuecomment-1708528606 is being done.

What I'm not sure about is if I've missed this, or if it's coming in another commit! 😅

sesponda replied:
> What wasn't clear to me in this implementation was when aggregate rules were evaluated on the results of other rules.

They are evaluated only if a violation rule reads the aggregates. This is the file I'm using for my local testing:

```
# METADATA
# description: Could not find a rule satisfying query pattern (some_package).public.export
package custom.regal.rules.acmecorp["require-export-rule"]

import future.keywords
import data.regal.result

# First, let's collect which files satisfy the rule.
aggregate contains entry if {
    input["package"].path[2].value == "public"
    some i
	input["rules"][i].head.name == "export"
	input["rules"][i].head.assign == true
	input["rules"][i].head.value.type == "array"
    entry := { "export_rule_found" : input.regal.file.name }
}

report contains violation if {
    not there_is_at_least_one_export_rule
    violation := result.fail(rego.metadata.chain(), {})
}

there_is_at_least_one_export_rule {
	input.aggregate[_].export_rule_found
}
```

Notes: 
* This could be simplified by simple checking if we have >0 aggregates. I wanted to check the negation as a way to see how one could have multiple aggregates collecting different data, and having different violation rules not interfering with each other.
* The aggregate record is an object, which makes it very flexible. One could abuse it, though. For example:

```
# Danger zone below - just to should the potential
aggregate contains entry if {
    # A way to collect all ASTs, providing the author understands the performance limits and risks.
    entry := { "AST" : input }
}
report contains violation if {
    input.aggregate[_].AST["package"].path[2].value == "public"
    violation := result.fail(rego.metadata.chain(), {})
}
```

charlieegan3 replied:
> They are evaluated only if a violation rule reads the aggregates

Ahh 🤦, gotcha! I knew there was something I was missing.

Getting a test in with that example make sense as a good next step to me.

> One could abuse it, though

I think we can live with that. We're not likely to include rules in upstream Regal of that format, but if people want to write and run their own rules that seems ok.

anderseknert replied:
From your example,

```rego
entry := { "export_rule_found" : input.regal.file.name }
```

It'd be good if we had an `result.aggregate` function added which always included `input.regal.file.name` in the returned object, so that we'd have a unique value per aggregate object. If we don't there's a risk they'll be munged as a single value when added to the set. But that's an optimization we can save for later too :)

charlieegan3 reviewed (commented):
Thanks for having a look into this, I think it's going to be a great feature.

sesponda commented:
Hi @anderseknert @charlieegan3 I've addressed the PR comment and pushed a new version, rebased from the recent changes in the main branch (newbie question: is it OK to rebase form main mid-PR, or you prefer a merge commit? - feel free to point me to best practices, etc.).

I have some test failures:
```
--- FAIL: TestLintRuleIgnoreFiles (0.23s)
    cli_test.go:236: expected two rules to be missing from reported violations, but there were 34 out of 38 violations reported
    cli_test.go:253: expected violation for rule "dubious-print-sprintf"
    cli_test.go:253: expected violation for rule "chained-rule-body"
--- FAIL: TestLintAllViolations (0.26s)
    cli_test.go:187: expected violation for rule "chained-rule-body"
    cli_test.go:187: expected violation for rule "dubious-print-sprintf"
--- FAIL: TestTestRegalBundledBundle (0.76s)
    cli_test.go:333: expected exit status 0, got 2
```

... which happen also in my updated main branch, so I'm not sure what's happening, I'll investigate. Maybe unrelated to my change? 

Next actions:
* If the recent update is OK -> continue with tests (once I sort out why I have other tests failing).
* Do you need help with documentation updates? How do you coordinate the release of a new build with docs? (probably I can learn that deeply studying the CI files but I'm not ashamed to recognise (a) I'm a newbie in Go (including Github CI), (b) I'm not afraid of asking dumb questions (there is a fine line between this and laziness, I know 😸 )

anderseknert commented:
@sesponda those are e2e tests failing, and the reason is they require a fresh binary built from the head of your changes (after you've rebased from main).

I recently added some code to make sure there _is_ a binary before starting the tests, but there's currently no way to know if it's stale or not. Perhaps it'd be best if we built the binary before starting those tests after all 🤔 Either way, I'll have to figure something out there. To solve the immediate problem, just `go build` and they should work.

(CC @srenatus as we've talked about this before, and you know the problems I ran into).

Help with docs would be great! I don't expect there'd have to be much to start with, and I can expand on them later, but something basic explaining how this works in the docs on custom rule development would be great!

anderseknert commented:
Continuing with some tests seems like a good next step to me, if @charlieegan3 considers his questions resolved.

Awesome work! 👏

charlieegan3 commented:
All good from me, thanks for taking the time to respond in detail 🙂

sesponda commented:
Hi @anderseknert @charlieegan3 thanks for your time so far. I started my work in spare time but later it became company-sponsored contribution (I work for Atlassian), so to keep things tidy I'm creating the same PR (same diff) from a different fork. Let's continue on https://github.com/StyraInc/regal/pull/323 (I've added the e2e tests)

sesponda commented:
Moved to: https://github.com/StyraInc/regal/pull/323
